### PR TITLE
Make the simple Node OAuth app more configurable

### PR DESCRIPTION
**Pull request by mjlumetta**
unless i'm mistaken, there isn't a good way to configure the simple Node app to run through the local OAuth proxy, since it uses the OpenID configuration [here](https://dev-api.va.gov/oauth2/.well-known/openid-configuration), and that points to `dev-api.va.gov` as the OAuth host. this PR adds a metadata file that points to the localhost proxy and allows users to use that configuration with a `--local` flag.

### CODE REVIEW

**Comment by dvogel** [comment]
This would definitely improve the local testing experience for the oauth-proxy. However this sample app is meant to be used by application developers as a test bed for interacting with our public offerings. I worry that including instructions and a sample config for running alongside the oauth-proxy will be misleading for that audience.

I think an alternative set of changes could eliminate a lot of the same pain without the potential confusion for others. When run locally, the oauth-proxy should provide it's own `openid-configuration` endpoint. This should allow us to:

1. Make this sample app support a command line option to override:
  a. the `openid-configuration` endpoint.
  b. the IDP.
2. Add a script to the oauth-proxy repo to run it with these options to point it at the local oauth-proxy on port 7100 and the IDP that redirects back to the local saml-proxy.
3. Move the directions you've added here to either `vets-saml-proxy/oauth-proxy/README.md` or into `vets-contrib`.

Thoughts?

**Comment by mjlumetta** [comment]
> However this sample app is meant to be used by application developers as a test bed for interacting with our public offerings.

i agree with what you're saying; it felt awkward to write 😅 but, that said, i think the root problem is that that primary purpose and the secondary purpose of testing in development are completely divergent. i think a better solution would be to copy this app to a new directory in `vets-saml-repo` called `test-clients` or something similar and build out our test environments there, which would then free up this repo to be unequivocally public-facing. as the program matures, both in terms of these specific components becoming more robust and in terms of having more API consumers using the platform, addressing both goals will become more and more unsustainable.

part of what influenced me toward this current iteration of a solution was that there are other instances in this repo (which i think have been developed more recently than this app?) where samples clearly favor program developers over external ones: https://github.com/department-of-veterans-affairs/vets-api-clients/blob/master/samples/oauth_react/util/default-config.js#L9. it makes way more sense that that isn't the pattern to follow, but it seems like those ought to be cleaned out in the future as well.

in any case, thoughts on that alternate direction to the alternate? (1) and (3) under your suggestions make plenty of sense to me, but (2) i'm not sure about because it creates a cross-repo dependency (right? unless i'm misunderstanding). that approach in (2) might be expedient in the short term, though. the con to my suggestion is that it would cause duplication in the short term, but it would separate concerns in a way that could easily be beneficial in the long run and allow each app to solve 1 problem well rather than trying to be very configurable for separate purposes.

**Comment by kalilsn** [comment]
@mjlumetta what do you see as the problems with keeping the sample apps in a separate repo from the saml/oauth proxies?

I would like us to keep them here and make the smaller changes Drew suggested because I think it's good for us to use the same sample apps that we're advertising to our users. If we started creating sample apps in a directory of `vets-saml-proxy`, then I could easily see us making changes to those and forgetting to update the user-facing ones.

I'd also just like to put in a vote for documenting this in the `vets-saml-proxy` repo rather than in `vets-contrib` (although adding a link in there would probably be good too). 

>part of what influenced me toward this current iteration of a solution was that there are other instances in this repo (which i think have been developed more recently than this app?) where samples clearly favor program developers over external ones: /samples/oauth_react/util/default-config.js@master#L9 . it makes way more sense that that isn't the pattern to follow, but it seems like those ought to be cleaned out in the future as well.

Yeah, this seems like a confusing mistake that should be fixed. I think it's fine for applications in this repo to be configurable enough for us to use them for local development, but all the defaults should be appropriate for our users to run them with minimal configuration.

**Comment by mjlumetta** [comment]
> I'd also just like to put in a vote for documenting this in the `vets-saml-proxy` repo rather than in `vets-contrib` (although adding a link in there would probably be good too).

sounds good to me. i would actually love for us to follow this pattern in general for technical documentation: use the `vets-contrib` Developer Process directory for indexing and overviewing various parts of the dev process but keep everything essential alongside the source code.

**Comment by mjlumetta** [comment]
> what do you see as the problems with keeping the sample apps in a separate repo from the saml/oauth proxies?

i don't think i'd say there's a problem with it, more so that i think there's a tradeoff worth discussing further. i think that when i said "the root problem", i was referring to the problem that initiated this PR, i.e. of not being able to run this app with the local OAuth proxy, as i understand it. but that may have also been overstating it.

> it's good for us to use the same sample apps that we're advertising to our users.

i don't necessarily disagree, but i don't necessarily agree either, so i'd like to toss out some thoughts and play out the string. first, can you make explicit what you see as the benefit of using the same apps for both purposes? i can think of one big benefit: if we use this app every time we update either of the proxies, that gives us confidence that the public-facing apps' integration with the authentication flow works, that we aren't giving our users a broken product. important. on the other hand, that would mean that we're committed to manually testing these apps on a pretty regular basis, and i'm not in love with that idea, though it def makes sense in the short term. if we could put in some work to write e2e tests for our sample apps (and i think this would be a great thing to invest in at some point!), then i don't know that it's so necessary to test manually frequently. probably pros and cons to having e2e tests in `vets-api-clients`, but i'll come back to that.

what are your thoughts on that? do you have other reasons you see as beneficial to that practice? i guess what i'm saying is that i can see a rationale for this approach, but i also think that we can solve that problem a different way. open to being convinced if there are reasons you're thinking of that i'm not. 

> If we started creating sample apps in a directory of `vets-saml-proxy`, then I could easily see us making changes to those and forgetting to update the user-facing ones.

i'm not sure i follow why we'd want to do this. being able to not have to do this is an assumption behind the approach i proposed. in addition to the e2e step from above, another small tradeoff i can see is the one between using the sample client frequently and having to keep both purposes in mind whenever we change this client. i don't know what future enhancements we'll need to make, if any, to test the proxies locally in the future. if this client is for both external users to learn how to integrate with OAuth APIs and for us to test the authentication flow, then we'll have to consider what's okay for external users to see every time we want to make something easier to test. if we don't think we'll ever have to make enhancements for testing locally, then this isn't really a concern... but otherwise, think this is worth considering. 

couple of other _really_ small reasons to separate these apps:

* on the opposite side of the testing enhancements consideration: this app is not a great demonstration of our OAuth APIs. i would think that as more users use this repo as a resource for learning how to build OAuth integrations, the product owners will want to showcase more of the OAuth APIs in this Node app. i'm actually a big fan of how dumb this app is for testing purposes because it is super uncomplicated to understand and wouldn't mind keeping it around for that reason.
* if i'm an external developer exploring this repo, i would like everything in the repo to be useful and/or comprehensible for me. some people are skimmers, and some people like to read everything. configuring the OpenID configuration might be plausible if they wanted to test like the prod endpoint or something (i guess?). can't see why they would need to know what the IdP is and don't know that all users of this repo will get that right away. for example, let's say a developer who has never developed an OAuth application and knows little to nothing about OAuth in general browsing this repo to learn how to do that. would they get right away that that's just an implementation detail on our end that's configured as part of the client, or is there a possibility that they would think it's something they would need to include in their client app? i don't know, but i think it's likely. we could add explanatory comments along the lines of "you don't need to do this in your client app", but that might be distracting or confusing in a way that is detrimental to the purpose of learning from the sample client. 

### Summary: Pros and Cons of Things

#### Diverging clients

**Pros**

- freedom to make enhancements to the testing app if there are things we'd like to do to make testing easier that might be too sensitive or too irrelevant for the public-facing repo
  - "do one thing well" principle
- e2e tests for `vets-api-clients` are probably worthwhile in their own right
- simpler for Lighthouse developers to get started working in `vets-saml-proxy` if the Node app ever changes
- confidence we're not providing confusing or distracting information to developers who are using this repo to learn OAuth, esp inexperienced ones
- can `docker-compose` the OAuth/SAML proxy test environment (i guess we can do this with Drew's idea too, but like, think this is a bit nicer)

**Cons**

- reasons to use the same app as our external users that i am not considering
- unforeseen problems with e2e tests idea?
- probably more work

#### Centralizing the sample client and the test client in the same app

**Pros**

- confidence of manually testing our sample app
- less work

**Cons**

- expose implementation details to external users learning from these clients and potentially distract from the educational purpose of the clients
- have to consider both purposes of the application whenever we make changes in the future, especially when making changes aimed at benefitting internal users
- relying on manual testing?

#### e2e tests for sample clients

**Pros**

- easier to test clients with automated tests
- sample code that external developers can use as a starting point if they would like to write e2e browser tests for their client apps 

**Cons**

- hard to automate e2e tests in a CI setting? for the same reasons that i'm saying we should consider using separate apps, we wouldn't want to put a jenkinsfile in this repo
- other considerations i'm missing?

### TLDR

that's my (way, way, way too long) line of thought on our options. i would like to move in this direction, but i am v open to being convinced if i am not considering potential problems to this approach or am overestimating the evidence for that approach. it's absolutely possible that i am way off base here.

**Comment by mjlumetta** [comment]
also, aren't we already doing the work to essentially write e2e tests with the token fetcher? as long as we're thinking about how to do automate stepping through the login flow... maybe we can re-use that logic here?

**Comment by beaugrantham-va** [comment]
Closing stale PRs.
